scripts/build-engine/src/index.ts

```diff
@@ -223,165 +196,313 @@ async function _doBuild ({
+    const [
+        { getPublicModules, getForceStandaloneModules },
+        { generateBuildTimeConstants },
+        { default: buildConfigFunction },
+    ] = await Promise.all([
+        import(`${libWhoAmI}/public-modules`) as Promise<typeof import('../../who-am-i/lib/public-modules')>,
+        import(`${libWhoAmI}/build-time-constants`) as Promise<typeof import('../../who-am-i/lib/build-time-constants')>,
+        import(`${libWhoAmI}/contextual-build-config`) as Promise<typeof import('../../who-am-i/lib/contextual-build-config')>,
+    ]);
 
-    const rpVirtualOptions: Record<string, string> = {};
-    const vmInternalConstants = getModuleSourceInternalConstants(options.buildTimeConstants);
-    console.debug(`Module source "internal-constants":\n${vmInternalConstants}`);
-    rpVirtualOptions['internal:constants'] = vmInternalConstants;
+    // The modules included in this build.
+    const includedPublicModules = await filterIncludedPublicModules();
 
-    const forceStandaloneModules = ['cc.wait-for-ammo-instantiation', 'cc.decorator'];
+    // Make up 'internal:constants'.
+    await makeUpInternalConstantsModule();
 
-    let rollupEntries: NonNullable<rollup.RollupOptions['input']> | undefined;
-    if (split) {
-        rollupEntries = { ...engineEntries };
-    } else {
-        rollupEntries = {
-            cc: 'cc',
-        };
-        const bundledModules = [];
-        for (const moduleName of Object.keys(engineEntries)) {
-            const moduleEntryFile = engineEntries[moduleName];
-            if (forceStandaloneModules.includes(moduleName)) {
-                rollupEntries[moduleName] = moduleEntryFile;
-            } else {
-                bundledModules.push(filePathToModuleRequest(moduleEntryFile));
-            }
-        }
+    // Build config.
+    const config = await getConfig();
+
+    // Rollup inputs.
+    const rollupInput = getRollupInput(includedPublicModules);
+
+    // Rollup plugins.
+    const rollupPlugins = await collectRollupPlugins();
+
+    const rollupOptions: rollup.InputOptions = {
+        input: rollupInput,
+        plugins: rollupPlugins,
+        cache: false,
+    };
+
+    const ammoJsAsmJsModule = await nodeResolveAsync('@cocos/ammo/builds/ammo.full.js');
+    const ammoJsWasmModule = await nodeResolveAsync('@cocos/ammo/builds/ammo.wasm.js');
+    if (ammoJsWasm === 'fallback') {
+        rpVirtualOptions['@cocos/ammo'] = `
+let ammo;
+let isWasm = false;
+if (typeof WebAssembly === 'undefined') {
+    ammo = await import('${filePathToModuleRequest(ammoJsAsmJsModule)}');
+} else {
+    ammo = await import('${filePathToModuleRequest(ammoJsWasmModule)}');
+    isWasm = true;
+}
+export default ammo.default;
+export { isWasm };
+`;
+    } else if (ammoJsWasm === true) {
+        rpVirtualOptions['@cocos/ammo'] = `
+import Ammo from '${filePathToModuleRequest(ammoJsWasmModule)}';
+export default Ammo;
+const isWasm = false;
+export { isWasm };
+`;
```

Comment:
In future I'm going to generalize the code here. But I have no idea now. @JayceLai